### ISSUE TITLE
JavadocParagraph: violate preceding P tag before block-level HTML tags

### ISSUE THREAD

**Issue by Zopsss** [checkstyle/checkstyle#15503]
Follow up of https://github.com/checkstyle/checkstyle/issues/15458

From https://github.com/checkstyle/checkstyle/pull/14901/files#r1622428372

https://checkstyle.org/checks/javadoc/javadocparagraph.html#JavadocParagraph

Investigated at : 
https://github.com/checkstyle/checkstyle/issues/15011#issuecomment-2174350170 (there are screenshots of javadoc with extra P tag)

[7.1.2 Paragraphs](https://google.github.io/styleguide/javaguide.html#s7.1.2-javadoc-paragraphs) rule of google java style guide states that there should not be preceding `<p>` tag for block tags:

> HTML tags for other block-level elements, such as `<ul>` or `<table>`, are not preceded with `<p>`.

Whole rule:
> 7.1.2 Paragraphs
One blank line—that is, a line containing only the aligned leading asterisk (*)—appears between paragraphs, and before the group of block tags if present. Each paragraph except the first has `<p>` immediately before the first word, with no space after it. HTML tags for other block-level elements, such as `<ul> or <table>`, are not preceded with ` <p>`.

We currently have no support to check this, there's also a message explaining this gap in our config at [coverage table](https://checkstyle-diff-reports.s3.us-east-2.amazonaws.com/6e451a8_2024210116/google_style.html#a7.1.2):

> We don't have support to check for the extra preceding p tag before block tag; there is known issue to cover it fully: [#15011](https://github.com/checkstyle/checkstyle/issues/15011)

At https://github.com/checkstyle/checkstyle/issues/15458#issuecomment-2282837296, we decided to upgrade [JavadocParagraph](https://checkstyle.org/checks/javadoc/javadocparagraph.html#JavadocParagraph)'s implementation and add support to check if a block-level HTML tag is preceded by `<p>`, if it is then give violation for it.

---

**How it should work?**

This is a list of block-level tags in HTML:

```
address
article
aside
blockquote
canvas
dd
div
dl
dt
fieldset
figcaption
figure
footer
form
h1
h2
h3
h4
h5
h6
header
hr
li
main
nav
noscript
ol
p
pre
section
table
tfoot
ul
video
```

source w3 schools: https://www.w3schools.com/html/html_blocks.asp

We should check if any of such block-level HTML tag is not preceded by `<p>`.

**Example:**

```java
/**
 * Some summary.
 * 
 * <p><h1>Testing...</h1></p> // should give violation extra P tag
 */
public class InvalidParagraphTagPosition {
  /**
   * Some summary.
   * 
   *<p>  // should give violation - extra P tag
   *  <ul>
   *    <p>
   *      <li>1</li> // should NOT give violation as there is not empty line before
   *    </p>
   *  </ul>
   *</p>
   *
   * <p><b>testing</b></p> // ok, inline HTML tag. Not a block-level tag
   */
  public void foo() {}

  /**
   *  Some summary.
   *
   * <p> // should give violation: extra P tag. One more possible: after P should be no space
   *  <table> 
   *  </table>
   * </p>
   */
  public void fooo() {}
}
```

**Config:**

```xml
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
          "-//Checkstyle//DTD Checkstyle Configuration 1.3//EN"
          "https://checkstyle.org/dtds/configuration_1_3.dtd">

<module name="Checker">
  <module name="TreeWalker">
    <module name="JavadocParagraph" />
  </module>
</module>
```

**Output:**

```
$ java -jar .\checkstyle-10.17.0-all.jar -c .\config.xml .\InvalidParagraphTagPosition.java
Starting audit...
Audit done.
```

**Expected output:**
```
$ java -jar .\checkstyle-10.17.0-all.jar -c .\config.xml  .\InvalidParagraphTagPosition.java
Starting audit...
[ERROR] .\InvalidParagraphTagPosition.java:2:7:  Block-level HTML tag "<h1>" should not be preceded with "<p>"
[ERROR] .\InvalidParagraphTagPosition.java:7:7:  Block-level HTML tag "<ul>" should not be preceded with "<p>"
[ERROR] .\InvalidParagraphTagPosition.java:21:7:  Block-level HTML tag "<table>" should not be preceded with "<p>"
Audit done.
Checkstyle ends with 4 errors.
```

The warning is given for the block-level HTML tag, explaining the tag is preceded by `<p>` which is not allowed.

---

**Additional context**

The coverage table website is generated by `GitHub, generate site`. We have not released new version so the live website does not contain explaining message, you won't be able to see it there.

**Comment by Zopsss** [comment]
referring to: https://github.com/checkstyle/checkstyle/issues/15458#issuecomment-2282281012

```
/**
 * Some Javadoc.
 *
 * <ul>Some sontent.</ul>
 *
 * <b>Some content.</b>
 */
 // violation 2 lines above 'P tag or block-level HTML elements should start paragraph'
```

Here, we should not give violation for `<b>` because it is an inline HTML tag. We should give violation only if the HTML tag is a block-level tag and it is preceded by `<p>`

**Comment by romani** [comment]
@rdiachenko , please review issue and place approval label if you agree.

**Comment by rdiachenko** [comment]
> ```
>    *<p>
>    *  <ul> // should give violation - extra P tag
>    *    <p>
>    *      <li>1</li> // should NOT give violation as there is not empty line before
>    *    </p>
>    *  </ul>
>    *</p>
> ```

`<li>` is a block element, I think we should give a violation here because of preceding `<p>`.

**Comment by romani** [comment]
@Zopsss , as you start this issue, please create quick update on Inputs to declare most of cases that come to mind and declaring by violations comments place where violation is expected. Send this PR for review, CI will be red is ok, at least we will align one more time on expected behavior. You meanwhile will start coding and will simply make CI pass, ideally without changes to Inputs.

**Comment by Zopsss** [comment]
@romani 

From example given in issue description: 

>  \* `<p>`
>  \*  `<table>` // should give violation: extra P tag. One more possible: after P should be no space

**"One more possible: after P should be no space"** this means that there should be no space immediately after P tag. We have a property for this [allowNewlineParagraph](https://checkstyle.org/checks/javadoc/javadocparagraph.html#Properties) which by defaults allows new line after P tag, if we decide not allow for newlines then we need to disable it.

From: https://github.com/checkstyle/checkstyle/issues/15503#issuecomment-2289878425
> Let's just do what they care is: after empty line there should be P tag or any other html block-level.

so do we need to disable allowNewlineParagraph or not as part of this issue?

**Comment by romani** [comment]
This property introduced at https://github.com/checkstyle/checkstyle/issues/1332

Examples of Checks in website are weird and unlikely to be valid. But it is unrelated problem.

We should not change this property, it should continue to allow new line, as some sort of lube wrap, not actually extra space.
As far as I remember, in html single space is rendering as space, all sequential are not rendering. When javadoc content in taken from javadoc comment, all new lines are removed by javadoc tool, so they are not a concern.

**Comment by romani** [comment]
@Zopsss , please send first PR with update for Inputs to define most use cases and behavior on them.
commit message `supplemental: tests for #15503`
lines with future violations should looks like `/* no violation until #15503*/` (attention to usage of multi-line comment)

We will merge this PR quickly, update for logic most likely will be in review longer.

**Comment by Zopsss** [comment]
https://github.com/checkstyle/checkstyle/pull/15680 PR sent @romani

**Comment by Zopsss** [comment]
I was trying to solve this issue locally, I found some bugs in [JavadocParagraph](https://checkstyle.org/checks/javadoc/javadocparagraph.html) which I would like to report before proceeding further.

```xml
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
          "-//Checkstyle//DTD Checkstyle Configuration 1.3//EN"
          "https://checkstyle.org/dtds/configuration_1_3.dtd">

<module name="Checker">
  <module name="TreeWalker">
    <module name="JavadocParagraph"/>
  </module>
</module>
```

```java
/**
 * Some summary.
 *
 * <p>  Some Javadoc.
 */
public class InvalidParagraphTagPosition {}
```

```
$ java -jar .\checkstyle-10.18.1-all.jar -c .\javadocparagraph_config.xml .\InvalidParagraphTagPosition.java
Starting audit...
[ERROR] C:\checkstyle testing\.\InvalidParagraphTagPosition.java:4: <p> tag should be placed immediately before the first word, with no space after. [JavadocParagraph]
Audit done.
```

We're facing a violation for ` * <p>  Some Javadoc.` which is correct but if I add `</p>` at the end of line then violation goes away

```java
/**
 * Some summary.
 *
 * <p>  Some Javadoc.</p>
 */
public class InvalidParagraphTagPosition {}
```

```
$ java -jar .\checkstyle-10.18.1-all.jar -c .\javadocparagraph_config.xml .\InvalidParagraphTagPosition.java
Starting audit...
Audit done.
```

This is happening because of different parsing of P tag when it is not closed and when it is closed.

AST of above example when `P` tag isn't closed ( `<p>  Some Javadoc.` ):

```
$ java -jar .\checkstyle-10.18.1-all.jar -J .\InvalidParagraphTagPosition.java
COMPILATION_UNIT -> COMPILATION_UNIT [6:0]
`--CLASS_DEF -> CLASS_DEF [6:0]
    |--MODIFIERS -> MODIFIERS [6:0]
    |   |--BLOCK_COMMENT_BEGIN -> /* [1:0]
    |   |   |--COMMENT_CONTENT -> *\r\n * Some summary.\r\n *\r\n * <p>  Some Javadoc.\r\n  [1:2]
    |   |   |   `--JAVADOC -> JAVADOC [1:3]
    |   |   |       |--NEWLINE -> \r\n [1:3]
    |   |   |       |--LEADING_ASTERISK ->  * [2:0]
    |   |   |       |--TEXT ->  Some summary. [2:2]
    |   |   |       |--NEWLINE -> \r\n [2:16]
    |   |   |       |--LEADING_ASTERISK ->  * [3:0]
    |   |   |       |--NEWLINE -> \r\n [3:2]
    |   |   |       |--LEADING_ASTERISK ->  * [4:0]
    |   |   |       |--TEXT ->   [4:2]
    |   |   |       |--HTML_ELEMENT -> HTML_ELEMENT [4:3]
    |   |   |       |   `--P_TAG_START -> P_TAG_START [4:3]
    |   |   |       |       |--START -> < [4:3]
    |   |   |       |       |--P_HTML_TAG_NAME -> p [4:4]
    |   |   |       |       `--END -> > [4:5]
    |   |   |       |--TEXT ->   Some Javadoc. [4:6]
    |   |   |       |--NEWLINE -> \r\n [4:21]
    |   |   |       |--TEXT ->   [5:0]
    |   |   |       `--EOF -> <EOF> [5:1]
    |   |   `--BLOCK_COMMENT_END -> */ [5:1]
    |   `--LITERAL_PUBLIC -> public [6:0]
    |--LITERAL_CLASS -> class [6:7]
    |--IDENT -> InvalidParagraphTagPosition [6:13]
    `--OBJBLOCK -> OBJBLOCK [6:41]
        |--LCURLY -> { [6:41]
        `--RCURLY -> } [6:42]
```

AST when `P` tag is closed ( `<p>  Some Javadoc.</p>` ):

```
java -jar .\checkstyle-10.18.1-all.jar -J .\InvalidParagraphTagPosition.java
COMPILATION_UNIT -> COMPILATION_UNIT [6:0]
`--CLASS_DEF -> CLASS_DEF [6:0]
    |--MODIFIERS -> MODIFIERS [6:0]
    |   |--BLOCK_COMMENT_BEGIN -> /* [1:0]
    |   |   |--COMMENT_CONTENT -> *\r\n * Some summary.\r\n *\r\n * <p>  Some Javadoc.</p>\r\n  [1:2]
    |   |   |   `--JAVADOC -> JAVADOC [1:3]
    |   |   |       |--NEWLINE -> \r\n [1:3]
    |   |   |       |--LEADING_ASTERISK ->  * [2:0]
    |   |   |       |--TEXT ->  Some summary. [2:2]
    |   |   |       |--NEWLINE -> \r\n [2:16]
    |   |   |       |--LEADING_ASTERISK ->  * [3:0]
    |   |   |       |--NEWLINE -> \r\n [3:2]
    |   |   |       |--LEADING_ASTERISK ->  * [4:0]
    |   |   |       |--TEXT ->   [4:2]
    |   |   |       |--HTML_ELEMENT -> HTML_ELEMENT [4:3]
    |   |   |       |   `--PARAGRAPH -> PARAGRAPH [4:3]
    |   |   |       |       |--P_TAG_START -> P_TAG_START [4:3]
    |   |   |       |       |   |--START -> < [4:3]
    |   |   |       |       |   |--P_HTML_TAG_NAME -> p [4:4]
    |   |   |       |       |   `--END -> > [4:5]
    |   |   |       |       |--TEXT ->   Some Javadoc. [4:6]
    |   |   |       |       `--P_TAG_END -> P_TAG_END [4:21]
    |   |   |       |           |--START -> < [4:21]
    |   |   |       |           |--SLASH -> / [4:22]
    |   |   |       |           |--P_HTML_TAG_NAME -> p [4:23]
    |   |   |       |           `--END -> > [4:24]
    |   |   |       |--NEWLINE -> \r\n [4:25]
    |   |   |       |--TEXT ->   [5:0]
    |   |   |       `--EOF -> <EOF> [5:1]
    |   |   `--BLOCK_COMMENT_END -> */ [5:1]
    |   `--LITERAL_PUBLIC -> public [6:0]
    |--LITERAL_CLASS -> class [6:7]
    |--IDENT -> InvalidParagraphTagPosition [6:13]
    `--OBJBLOCK -> OBJBLOCK [6:41]
        |--LCURLY -> { [6:41]
        `--RCURLY -> } [6:42]
```

Notice AST when `P` isn't closed:

```
    |   |   |       |--HTML_ELEMENT -> HTML_ELEMENT [4:3]
    |   |   |       |   `--P_TAG_START -> P_TAG_START [4:3]
```

and when `P` is closed:

```
    |   |   |       |--HTML_ELEMENT -> HTML_ELEMENT [4:3]
    |   |   |       |   `--PARAGRAPH -> PARAGRAPH [4:3]
    |   |   |       |       |--P_TAG_START -> P_TAG_START [4:3]
```

when `P` is fully closed then generated AST contains `--PARAGRAPH` token. 

What we have done in Check's implementation is:

https://github.com/checkstyle/checkstyle/blob/d07072b179330e40357c7c8852883af815d6978c/src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/JavadocParagraphCheck.java#L147-L156

we're checking if after `HTML_ELEMENT ` is there an opening P tag i.e `P_TAG_START`, if yes then move further.

This case is only valid when we don't close the `P` tag, if we close the `P` tag then a new token appears between HTML_ELEMENT & P_TAG_START so the check ignores that paragraph and we don't get the expected violation as shown in above example.

This is a critical bug IMO so my question is, how should we proceed with the current issue? should we just continue with this incorrect implementation or we should first fix the implementation and then move forward. 

Fixing the implementation will take some time and it will be out of scope of this issue, we will need to open another issue for it. @romani @rdiachenko

**Comment by romani** [comment]
this is new bug, please open issue on it.
I think it will be better to fix this new defect first, it should be easy fix, and it will simply further fix.

**Comment by Zopsss** [comment]
Issue opened: https://github.com/checkstyle/checkstyle/issues/15685 @romani
